Add iterator node

jrmi opened the pull request:
### What is in this PR

This is a big baby and it's not finished. This MR introduces the new graph dimension with container nodes like the iterator tha iterates over item of an array or like the Repeat node tha repeat the same sequence of node a defined amount of time (This last node is not included in the MR but should be relatively easy to add).

It also completely refactor the way the node graph is supported. Now the graph is a JSONB property of the workflow. A graph handler contains all the complexity regarding the graph manipulation and other layers are simpler. Read the backend `graph_handler.py` file doc for more information. Let me know if you need more details about it.

Additionnaly, the workflow starts with no trigger and the user can select one from the menu.

I've also fixed the icons of the nodes and the returned values by the local baserow triggers that wern't returned the right data shape as they are marked as `returns_list`. The list rows action was broken because of that.

For the iterator node, you can select an array and not only leaf in the data provider.

### Known limitations

#### To fix in this PR
- For now I haven't fully supported all edge cases but they are a bunch of TODOs and fake exception in the code to do that.
- The graph handlers aren't directly tested yet but they are tested through the different actions.
- There is no data migration yet

#### Will fix after
- When we guess the data schema from the data generated by a local baserow node we see the technical field nabes and not the human field names so it's not really usable as is. Will be fixed in a follow up.

### How to test this PR

- Test all operations: creation, deletion, move, replace. In iterator and after routers.
- Undo/redo/undo/redo for all operations
- These operations need to be tested with router nodes and iterator nodes
- Try to create a complex graph with iterator and router and try to test run/simulate/publish and check all permutations are working properly.
- Duplicate a workflow and check the duplicated graph

### Checklist

- [ ] A changelog entry has been added to `changelog/entries/unreleased` using `changelog/src/changelog.py`
- [ ] New/updated **Premium/Enterprise features** are separated correctly in the premium or enterprise folder
- [ ] The latest **Chrome and Firefox** have been used to test any new frontend features
- [ ] [Documentation](https://github.com/baserow/baserow/blob/master/docs/index.md) has been updated
- [ ] [Quality Standards](https://github.com/baserow/baserow/blob/master/CONTRIBUTING.md#quality-standards) are met
- [ ] **Performance**: tables are still fast with 100k+ rows, 100+ field tables
- [ ] The [redoc API pages](https://api.baserow.io/api/redoc/) have been updated for any REST API changes
- [ ] Our [custom API docs](https://github.com/baserow/baserow/blob/master/web-frontend/modules/database/pages/APIDocsDatabase.vue) are updated for changes to endpoints accessed via API tokens
- [ ] The UI/UX has been updated following the [UI Style Guide](https://baserow.io/style-guide)
- [ ] Security impact of change has been considered

Closes #3781

paljort commented on backend/src/baserow/contrib/automation/api/nodes/serializers.py:
```diff
@@ -113,12 +95,17 @@ class ReplaceAutomationNodeSerializer(serializers.Serializer):
         required=False,
-        help_text="The ID of the node that should be before the moved node.",
+        help_text="The target position node.",
+    )
```
iiuc, the three fields in this serializer defines the destination that a node is being moved to, and the `position_node_id` is the reference point. If so, I'm wondering if `reference_node_id` or `relative_node_id` is a clearer name for this?

Maybe the help could say "The node will be moved relative to this node ID."

paljort reviewed (commented):
It is indeed a big one 😅 .

I'm leaving some early feedback as I can't proceed testing in the frontend. In the meantime, I'll continue reviewing the code.

paljort reviewed (commented):
Great work @jrmi - I think the refactor does seem to make the logic more localized 🚀 

I've gone through most of the code changes, and I think it makes sense. I'll take a closer look at the graph logic tomorrow. If the crash I mentioned earlier is fixed, I'll also test the frontend more thoroughly tomorrow.

jrmi commented:
> I had an existing workflow,

As I said in the description, the migration of existing workflow is not yet implemented that's probably why it fails.

jrmi commented on backend/src/baserow/contrib/automation/api/nodes/serializers.py:
```diff
@@ -64,23 +47,30 @@ class CreateAutomationNodeSerializer(serializers.ModelSerializer):
         required=False,
-        help_text="If provided, creates the node after this given id.",
+        allow_blank=True,
+        help_text="The unique ID of the branch this node is an output for.",
```
Updated (not same wording but close enough)

jrmi commented on backend/src/baserow/contrib/automation/api/nodes/views.py:
```diff
@@ -111,7 +114,7 @@ def get_permissions(self):
         {
             AutomationWorkflowDoesNotExist: ERROR_AUTOMATION_WORKFLOW_DOES_NOT_EXIST,
-            AutomationNodeBeforeInvalid: ERROR_AUTOMATION_NODE_BEFORE_INVALID,
+            AutomationNodePositionNodeInvalid: ERROR_AUTOMATION_NODE_POSITION_NODE_INVALID,
```
Replaced with reference node

jrmi commented on backend/src/baserow/contrib/integrations/core/service_types.py:
```diff
@@ -1585,3 +1589,88 @@ def export_prepared_values(
+        self,
+        data: Any,
+    ) -> DispatchResult:
+        return DispatchResult(data)
```
Done.

jrmi commented:
> > I had an existing workflow,
> 
> As I said in the description, the migration of existing workflow is not yet implemented that's probably why it fails.

There is no a forward migration.

paljort commented:
The Automation Builder guided tour is broken on steps 2, 3, and 4:
```
Uncaught (in promise) TypeError: Cannot read properties of null (reading 'top')
```

This is because the design has changed in this MR (e.g. there is no default trigger). To reproduce:
- Clear the value for your user profile's completed_guided_tours field. You can do this in baserow shell with e.g. `UserProfile.objects.filter(user__email="tsering@baserow.io").update(completed_guided_tours=[])`
- Create a new Automation app

When you reach steps 2-4 in the guide tour, you'll notice the error in the console.

paljort commented:
I created an Iterator node, and added a child action node (e.g. "Rows are updated"). In the child node's service, when I click on a field, the data explorer shows the Iterator node twice:
> <img width="321" height="244" alt="image" src="https://github.com/user-attachments/assets/8da073e4-2139-48a6-88e1-7eec307f993b" />

I'm wondering if it makes sense to show the Iterator node again under the "Previous node" section? I guess the Iterator node is semantically a previous node, but I think removing this would make it less confusing. If the automation designer wants to pick an arbitrary row, they could just pick from the "Rows are updated" node.

Also, I'm wondering why the "Rows are updated" node allows you to pick any row. Shouldn't it only allow you to pick the row that was actually updated?

paljort commented on backend/src/baserow/contrib/automation/api/nodes/serializers.py:
```diff
@@ -64,23 +48,32 @@ class CreateAutomationNodeSerializer(serializers.ModelSerializer):
         required=False,
-        help_text="If provided, creates the node after this given id.",
+        allow_blank=True,
+        help_text="The position of the new node relatively to the reference node.",
```
```suggestion
        help_text="The position of the new node relative to the reference node.",
```

paljort commented:
The message about testing the previous node looks a bit weird with the hyphens:
> <img width="353" height="118" alt="image" src="https://github.com/user-attachments/assets/407b19d8-b15d-4fa8-9805-955e974e609a" />

Maybe we can enclose the previous node's name in quotes instead? E.g. `The previous node 'Rows are deleted' must be configured.`

jrmi commented:
> I think there is a regression from your last commit. It seems that I can't get a basic Iterator node to work now - I'm pretty sure it was working in the last commit 🤔

Thanks for this second review. My answers:

> I think it's because in both places where this data-highlight is used (WorkflowEditor.vue and WorkflowNode.vue), they assume that a node exists. But since we no longer create a default trigger node, this selector doesn't exist.

Actually I fixed this one already after your last comment. This selector is in the `WorflowEditor.vue` file now but in the next step `data-highlight="automation-node-sidepanel"` is missing as it's the side panel and we don't show the side panel anymore because we don't have any node selected (neither created). I have changed the guided tour again to remove old steps.

> Currently we say "trigger or action", but only a trigger can be selected.
> In general, I think this step 4 should be re-written to be more accurate.

Done, what do you think?

> Should we treat the Source like how we treat a Create Row node that doesn't have a Table defined? I.e. the node won't let you test a Create Row node until you set the Table. Maybe we should do the same for an Iterator node: don't allow testing it until a Source is defined?

Actually it was the case but after the formula refactoring the check was broken. Good catch. Now it should be as you described.

> Currently, if you don't set the Source and just test it, the sample data is set to []. The subsequent child node cannot see the Current Iteration in the data explorer.

And this case can happen even with a valid formula. Something is missing in the data explorer to say that a node doesn't have a schema and why (unstested, no schema at all, not configured, ...). I would say this is out of scope of this MR and it could conflict with existing work on another branch so if you don't mind I'll fix that in a follow up MR.

> I found that the published workflow crashes upon running. This is strange, because I think it was working in the last commit.

A bunch of tests were failing, I should have checked that, the import/export of the graph was broken... Sorry about that, it's fixed now :wink:

paljort commented on backend/src/baserow/contrib/automation/api/nodes/errors.py:
```diff
@@ -1,7 +1,7 @@
 from rest_framework.status import HTTP_400_BAD_REQUEST, HTTP_404_NOT_FOUND
 
```
Do we want to allow Iterator nodes to be movable? If not, I think we should make it similar to the trigger node, in that the node cannot be draggable:
> https://github.com/user-attachments/assets/00845ebe-d1b6-4e09-9c69-cd0a3b4345b4

paljort reviewed (approved):
I think this is working well in various scenarios that I tested 🚀
